### DISCUSSION TITLE
Update another table during registration, register event

### DISCUSSION THREAD

**Discussion by mauriceb77** [codeigniter4/shield#472] [Q&A] [answered]
Hello,

My site only allows user to register when they are referred to the site and they enter a valid registration code, I have this working fine. However, I need to update my user referral table with the newly created user ID when the user is registering. I am using the "register" event triggered by Shield's RegisterController to call a custom class::method but as the user is not yet logged in, I am not sure how to get the current user ID.

Is using the event the correct approach and if so, how can I get the current user ID without using auth()->id()?

Thanks,
-mb

**Comment by datamweb** [comment]
You can use the following explanations. 
https://github.com/codeigniter4/shield/blob/develop/docs/events.md#register
Have you read this section? If yes, what is unclear?

**Comment by MGatner** [comment, marked as answer]
In addition to @datamweb's suggestion the piece you might be missing is that the new user is passed into the Register event as a parameter, so when you set up your listener it can receive the object:
```php
Events::on('register', static function ($user) {
    echo $user->id;
});
```

**Comment by jozefrebjak** [comment]
> Hello,
> 
> 
> 
> My site only allows user to register when they are referred to the site and they enter a valid registration code, I have this working fine. However, I need to update my user referral table with the newly created user ID when the user is registering. I am using the "register" event triggered by Shield's RegisterController to call a custom class::method but as the user is not yet logged in, I am not sure how to get the current user ID.
> 
> 
> 
> Is using the event the correct approach and if so, how can I get the current user ID without using auth()->id()?
> 
> 
> 
> Thanks,
> 
> -mb

Can you share what you extend? I would like to implement something similar

**Reply by mauriceb77** [reply]
@jozefrebjak

Below is what I implemented.

I created my own register view based on Shield's view and update the Auth config to use my view.

**Config/Validation: Added rule to validate the registration code against the registering email. **

```
use \App\Validation\RegisterRules;

    public $ruleSets = [
        ....
        RegisterRules::class,
    ];

public $registration = 
    [
        'rcode'    => [ 'label' => 'Registration Code',
                       'rules'  => 'required|max_length[50]|validate_code[rcode,email]'],
```

```
namespace App\Validation;

class RegisterRules
{
    public function validate_code($str, string $fields, array $data): bool
    {
        $UserModel = new \App\Models\User_model();

        if($UserModel->validate_registration_code($data['email'],$data['rcode']))
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}
```


**Config/Events: Added call to registration code redemption function on the RegisterController's "register" event.**

```
Events::on('register', function($user)
            {
                $registration = new \App\Libraries\Registration();
                $registration->redeem_referral($user);
            });
```

In my redeem_referral method, I update my referral table with the new user ID.

```
$UserModel = new \App\Models\User_model();

      $data = $UserModel->get_referral_key($user->id);
      if($UserModel->redeem_referral($data['recipient_email'],$data['code'],$user->id))
      {
....
      }
```
